compiler/rustc_driver_impl/src/signal_handler.rs

```diff
@@ -35,20 +35,18 @@ macro raw_errln($tokens:tt) {
 }
 
 /// Signal handler installed for SIGSEGV
-// FIXME(static_mut_refs): Do not allow `static_mut_refs` lint
-#[allow(static_mut_refs)]
 extern "C" fn print_stack_trace(_: libc::c_int) {
     const MAX_FRAMES: usize = 256;
     // Reserve data segment so we don't have to malloc in a signal handler, which might fail
     // in incredibly undesirable and unexpected ways due to e.g. the allocator deadlocking
     static mut STACK_TRACE: [*mut libc::c_void; MAX_FRAMES] = [ptr::null_mut(); MAX_FRAMES];
     let stack = unsafe {
         // Collect return addresses
-        let depth = libc::backtrace(STACK_TRACE.as_mut_ptr(), MAX_FRAMES as i32);
+        let depth = libc::backtrace(&raw mut STACK_TRACE as _, MAX_FRAMES as i32);
         if depth == 0 {
             return;
         }
-        &STACK_TRACE.as_slice()[0..(depth as _)]
+        slice::from_raw_parts(&raw const STACK_TRACE as _, depth as _)
```

Comment:
I feel like neither of these fixes the underlying issue that we likely shouldn't be using a static mut here... It just side-steps the lint by using raw pointers which for some reason we don't lint on.

Reply:
> by using raw pointers which for some reason we don't lint on.

We don't lint against raw pointers because that's often what we're trying to push people toward (among other options).  We lint against references because it's easier to get that wrong without realizing.  E.g., it's fine to have aliasing pointers to some data as long as you're careful to not race the accesses, but it's never fine to have aliasing live mutable references.  But taking potentially long-lived references to mutable statics makes it really easy to do that accidentally.

To make this particular case safer, I'd probably move `STACK_TRACE` into that `let stack = unsafe { .. }` block.  Then it can't be accessed later while the reference to it via that slice is live, which is the main danger.

Probably I'd also expect this function to be `unsafe extern "C" fn ..` since there's a safety requirement here that must be upheld by callers (the function is not reentrant).  Not sure if there's some practical reason preventing that here.